junit-platform-engine/src/main/java/org/junit/platform/engine/support/discovery/ResourceContainerSelectorResolver.java

```diff
@@ -0,0 +1,88 @@
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.engine.support.discovery;
+
+import static java.util.stream.Collectors.toSet;
+import static org.junit.platform.commons.support.ReflectionSupport.findAllResourcesInClasspathRoot;
+import static org.junit.platform.commons.support.ReflectionSupport.findAllResourcesInPackage;
+import static org.junit.platform.commons.util.ReflectionUtils.findAllResourcesInModule;
+import static org.junit.platform.engine.support.discovery.SelectorResolver.Resolution.selectors;
+import static org.junit.platform.engine.support.discovery.SelectorResolver.Resolution.unresolved;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import org.junit.platform.commons.support.Resource;
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.discovery.ModuleSelector;
+import org.junit.platform.engine.discovery.PackageSelector;
+
+/**
+ * @since 1.11
+ */
+class ResourceContainerSelectorResolver implements SelectorResolver {
+	private static final char CLASSPATH_RESOURCE_PATH_SEPARATOR = '/';
+	private static final char PACKAGE_SEPARATOR_CHAR = '.';
+	public static final String DEFAULT_PACKAGE_NAME = "";
+	private final Predicate<Resource> resourceFilter;
+	private final Predicate<String> resourceNameFilter;
+
+	ResourceContainerSelectorResolver(Predicate<Resource> resourceFilter, Predicate<String> resourcePackageFilter) {
+		this.resourceFilter = resourceFilter;
+		this.resourceNameFilter = adaptPackageFilter(resourcePackageFilter);
+	}
+
+	/**
+	 * A package filter is written to test {@code .} separated package names.
+	 * Resources however have {@code /} separated paths. By rewriting the path
+	 * of the resource into a package name, we can make the package filter work.
+	 */
+	private static Predicate<String> adaptPackageFilter(Predicate<String> packageFilter) {
+		return resource -> packageFilter.test(packageName(resource));
+	}
+
+	private static String packageName(String classpathResourceName) {
+		int lastIndexOf = classpathResourceName.lastIndexOf(CLASSPATH_RESOURCE_PATH_SEPARATOR);
+		if (lastIndexOf < 0) {
+			return DEFAULT_PACKAGE_NAME;
+		}
+		// classpath resource names do not start with /
+		String resourcePackagePath = classpathResourceName.substring(0, lastIndexOf);
+		return resourcePackagePath.replace(CLASSPATH_RESOURCE_PATH_SEPARATOR, PACKAGE_SEPARATOR_CHAR);
+	}
+
+	@Override
+	public Resolution resolve(ClasspathRootSelector selector, Context context) {
+		return resourceSelectors(
+			findAllResourcesInClasspathRoot(selector.getClasspathRoot(), resourceFilter, resourceNameFilter));
+	}
+
+	@Override
+	public Resolution resolve(ModuleSelector selector, Context context) {
+		return resourceSelectors(
+			findAllResourcesInModule(selector.getModuleName(), resourceFilter, resourceNameFilter));
```

Comment:
Currently, I'm assuming that all resources are class path resources, even if they're on the module path. Is that a valid assumption?

Reply:
I think that's reasonable. @sormuras WDYT?

Reply:
Resources are resources - running on Java 9+ all of them are in modules. Some of those modules are unnamend (loaded via the class-path), some are named (loaded via the module-path). So, maybe a bit late in this PR, but dropping "Classpath" from type and method names might be suited. For example `getClasspathResource` should be named `getResource` and `selectClasspathResource(Resource)` should be named `selectResource(Resource)`.

When loading resources in a named module the access is guarded by the module graph configuration supplied by the runtime. So yes, the implementation proposed should work as-is. A test confirming that would be nice to have, though.

Reply:
Since there's already `selectClasspathResource`, I wouldn't introduce a new name.

@mpkorstanje We'll discuss internally and will let you know before asking you to make any changes.

Reply:
**Team decision:** Stick with `DiscoverySelectors.selectClasspathResource` but document that the module path will also be checked.

Reply:
Resolved 27c367113cfb96b7663c737120f8e9aad3b9fdad